controllers/pac_pipelinerun_pruner_controller.go

```diff
@@ -0,0 +1,126 @@
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/go-logr/logr"
+
+	appstudiov1alpha1 "github.com/redhat-appstudio/application-api/api/v1alpha1"
+	tektonapi "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+)
+
+// PaCPipelineRunPrunerReconciler watches AppStudio Component object in order to clean up
+// running PipelineRuns created by Pipeline-as-Code when the Component gets deleted.
+type PaCPipelineRunPrunerReconciler struct {
+	Client        client.Client
+	Scheme        *runtime.Scheme
+	Log           logr.Logger
+	EventRecorder record.EventRecorder
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PaCPipelineRunPrunerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appstudiov1alpha1.Component{}, builder.WithPredicates(predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				return false
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return false
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return true
+			},
+			GenericFunc: func(e event.GenericEvent) bool {
+				return false
+			},
+		})).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=tekton.dev,resources=pipelineruns,verbs=get;list;watch;delete;deletecollection
+
+func (r *PaCPipelineRunPrunerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("PaCPipelineRunPruner", req.NamespacedName)
+
+	var component appstudiov1alpha1.Component
+	err := r.Client.Get(ctx, req.NamespacedName, &component)
+	if err == nil {
+		// The Component has been recreated.
+		log.Info("Component exists, nothing to do.")
+		return ctrl.Result{}, nil
+	}
+	if !errors.IsNotFound(err) {
+		log.Error(err, "failed to check Component existance")
+		return ctrl.Result{}, err
+	}
+
+	if err := r.PrunePipelineRuns(ctx, req); err != nil {
+		log.Error(err, "failed to prune PipelineRuns for the Component")
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// PrunePipelineRuns deletes PipelineRuns, if any, assocoated with the given Component.
+func (r *PaCPipelineRunPrunerReconciler) PrunePipelineRuns(ctx context.Context, req ctrl.Request) error {
+	log := r.Log.WithValues("PaCPipelineRunPruner", req.NamespacedName)
+
+	componentPipelineRunsRequirement, err := labels.NewRequirement(ComponentNameLabelName, selection.Equals, []string{req.Name})
+	if err != nil {
+		return err
+	}
+	componentPipelineRunsSelector := labels.NewSelector().Add(*componentPipelineRunsRequirement)
+	componentPipelineRunsListOptions := client.ListOptions{
+		LabelSelector: componentPipelineRunsSelector,
+		Namespace:     req.Namespace,
+	}
+
+	componentPipelineRunsList := &tektonapi.PipelineRunList{}
+	if err := r.Client.List(ctx, componentPipelineRunsList, &componentPipelineRunsListOptions); err != nil {
```

Comment:
I don't think the list prior to the delete provides you any real value (iirc the delete with the list option does a list under the covers for you anyway)

Just do the delete

Reply:
Thank you, Gabe, for the comment.
I do list prior to the delete to log information about cleaning for the Component (not needed or number of pipelines cleaned). Otherwise, we will not have any logs from the controller other than a Component was processed. The list operation is cheap because we read it from the cache (and the cache is customized, so we have only related to AppStudio pipelines there). So, to me the question is whether we need the logs or not.

Reply:
I personally don't think a log including the deleted items is worth it, even with the fact that you most likely will be hitting the cache on the list.

Code is much more concise with just the delete.

Just log prior to the delete stating you are deleting based on the component / label etc.